tests/integration/components/bs-button-test.js

```diff
@@ -9,11 +9,44 @@ import hbs from 'htmlbars-inline-precompile';
 import setupNoDeprecations from '../../helpers/setup-no-deprecations';
 import a11yAudit from 'ember-a11y-testing/test-support/audit';
 import sinon from 'sinon';
+/* global QUnit */
 
 module('Integration | Component | bs-button', function (hooks) {
   setupRenderingTest(hooks);
   setupNoDeprecations(hooks);
 
+  hooks.beforeEach(function (assert) {
+    /*
+     * The custom assertions `rejectsErrorNotGoinghThroughEmberOnError` allows to test
+     * functionality, which rejects intentionally. It serves a similar use case as
+     * `assert.throws` and `setupOnerror` utility provided by `@ember/test-helpers`.
+     * Sadly for some use cases none of them is working.
```

Comment:
Can you explain under which specific circumstances this does not work?

Reply:
`assert.throws` can only be used for synchronous functions. Test helpers like `click`, `visit`, `fillIn` provided by `@ember/test-helpers` are async. So that can not be used at all. But to be honest, I think I wanted to mention `assert.rejects` here. :see_no_evil: I addressed this in 4f036bc826b6cfc2c1ce38d00c2e720399df7f3c.

Test helpers like `click` and `fillIn` may cause application code to run, which throws an unhandled error or causes an unhandled rejection. But they will resolve even in that case. So `assert.rejects(async () => await click('button'))` will not work because `click('button')` will resolve even if it causes an unhandled rejection.

`setupOnerror` only works for errors, which are handled by `Ember.onerror`. It is only a nice helper about mocking `Ember.onerror` in tests.

Some errors don't go through `Ember.onerror`. This is discussed here: https://github.com/emberjs/ember-qunit/issues/592 To be honest I still don't understand the why. For me it sounds like a bug. :shrug: And I don't know for sure in which cases an error (or rejection) is not handled by `Ember.onerror`. Especially I don't understand why we aren't facing the same issues in the tests for `@onSubmit` handler of `<BsForm>`. :confused: 

When investigating (and before noticing that issue in ember-qunit) I created this playground to reproduce: https://github.com/jelhan/reproduce-component-event-handler-throws-testing-issues/compare/b95001721b9b027c444926db168975ed846fdac3...master

Reply:
Ok, thanks for the explanation!